prereise/cli/data_sources.py

```diff
@@ -0,0 +1,67 @@
+from powersimdata.input.grid import Grid
+from prereise.gather.winddata.rap import rap
+from prereise.cli.helpers import validate_date, validate_file_path
+
+
+class DataSource:
+    def __init__(self):
+        raise Exception("This is an abstract class, please do not use directly")
+
+    def add_to_download_parser(self, parser):
+        print("in add_to_download_parser")
+        subparsers = parser.add_subparsers()
+        subparser = subparsers.add_parser(self.command_name, help=self.command_help)
+        for argument in self.extract_arguments:
+            command = argument.pop("command_flags")
+            subparser.add_argument(*command, **argument)
+        subparser.set_defaults(func=self.extract)
+
+
+class WindData(DataSource):
+    def __init__(self):
+        self.command_name = "wind_data"
+        self.command_help = "this is the winddata help message"
+        self.extract_arguments = [
+            {
+                "command_flags": ["--region", "-r"],
+                "required": True,
+                "choices": ["Texas", "Eastern", "Western"],
+                "type": str,
+                "action": "append",
+            }, {
+                "command_flags": ["--start_date", "-sd"],
+                "required": True,
+                "type": validate_date,
+            }, {
+                "command_flags": ["--end_date", "-ed"],
+                "required": True,
+                "type": validate_date,
+            }, {
+                "command_flags": ["--file_path", "-fp"],
+                "required": True,
+                "type": validate_file_path,
+            }
+        ]
+
+    def extract(self, region, start_date, end_date, file_path, **kwargs):
+        regions = list(set(region))
+        assert start_date <= end_date
+
+        grid = Grid(regions)
+        wind_farms = grid.plant.groupby("type").get_group("wind")
+        data, missing = rap.retrieve_data(
```

Comment:
Any advice on how `missing` should be handled?

Reply:
I don't have much to add to that summary, but thinking the simplest way to handle it is just output the number of missing files, or the percentage, and let the user decide afterward whether the data available is sufficient. In the [rap_demo](https://github.com/Breakthrough-Energy/PreREISE/blob/develop/prereise/gather/winddata/rap/demo/rap_demo.ipynb) for example, that's all we really do, other than plotting the results which shows where the gaps are. 

Actually now that I think about it, I'd expect if the missing data is all sequential (so there is a large gap), it would be harder to impute the data. But there isn't an obvious threshold, so still think handling it programmatically might not be worth it. In this case though, maybe printing a full list of which days are missing in addition to the percentage might give the user more context. Wouldn't get more complex than this since most people would probably just rerun it until it works.